File: scala/largest-series-product/src/main/scala/Series.scala
object Series {

  type DigitString = String
  type Digits = Seq[Int]
  type Slizes = Seq[Digits]
  type Product = Int

  def digits(digitStr: DigitString): Digits =
    digitStr map (_.asDigit)

  private val NoSlices: Slizes = Seq()

  def slices(size: Int, digitStr: DigitString): Option[Slizes] = {
    val digitsSliding: DigitString => Slizes =
      digits(_) sliding size toSeq

    Option(digitStr)
      .filter (_.length >= size)
      .map (digitsSliding)
  }

  def largestProduct(size: Int, digitStr: DigitString): Option[Product] = {
    def getLargestProduct(slices: Slizes): Product =
      slices map (_.product) max

    if (size == 0)
      Some(1)
    else
      slices(size, digitStr) map getLargestProduct
  }
}

File: haskell/largest-series-product/Series.hs
module Series (largestProduct) where

import Data.Char (digitToInt)
import Data.List (tails)

type Size = Int
type DigitString = String
type Digits = [Int]
type Slizes = [Digits]
type Product = Int

slices :: Size -> DigitString -> Slizes
slices size digitString =
  filter ((== size) . length) allSlices
  where
    allSlices = map (take size) $ tails digits
    digits = map digitToInt digitString

largestProduct :: Size -> DigitString -> Maybe Product
largestProduct s d
  | null theSlices = Nothing
  | otherwise = Just $ maximum $ map product theSlices
  where
    theSlices = slices s d
